### PR TITLE
chore: migrate to TS & upgrade NextJS to v14

### PR DESCRIPTION
**Pull request by dhararaghu777**
## Summary by Sourcery

Migrate the project from Next.js 13 with JavaScript to Next.js 14 with TypeScript, introducing type safety and improving code maintainability. Add new components for cookie consent and service worker registration. Refactor existing components to use TypeScript and enhance the filtering functionality in the Contributors component. Update build and CI configurations to support TypeScript and improve code quality. Remove obsolete files and update documentation to reflect the changes.

New Features:
- Introduce TypeScript to the project by converting JavaScript files to TypeScript, enhancing type safety and code maintainability.
- Add a new CookieConsent component to manage user consent for cookies, improving compliance with privacy regulations.
- Implement a ServiceWorkerRegistration component to register service workers, enhancing offline capabilities and performance.
- Introduce a new layout structure using Next.js's app directory, improving the organization and scalability of the application.

Enhancements:
- Refactor components to use React Functional Components with TypeScript, improving code readability and type safety.
- Update the SEO configuration to use Next.js's Metadata type, enhancing SEO capabilities and type safety.
- Improve the styling and responsiveness of components by updating CSS and Tailwind configurations.
- Enhance the filtering functionality in the Contributors component by adding new filter options and improving the user interface.

Build:
- Add a TypeScript configuration file (tsconfig.json) to set up TypeScript in the project.
- Introduce an ESLint configuration file (eslint.config.js) to enforce coding standards and improve code quality.
- Update the Tailwind CSS configuration to use the new content paths and remove deprecated options.

CI:
- Remove old ESLint configuration files and replace them with a new centralized configuration, streamlining the linting process.

Documentation:
- Update documentation to reflect the migration from JavaScript to TypeScript, providing guidance on the new codebase structure and conventions.

Tests:
- Add type definitions and interfaces to improve test coverage and ensure type safety in test files.

Chores:
- Remove obsolete JavaScript files and components that have been replaced by their TypeScript counterparts, cleaning up the codebase.

### CODE REVIEW

**Comment by uniqueg** [comment]
@anuragxxd, @JaeAeich and/or @SalihuDickson - can you please review this PR? :pray:

**Comment by JaeAeich** [comment]
Wouldn't it make more sense to use elixir's eslint config defines in cloud component? @anuragxxd

File: src/app/contributors/page.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const fetchContributors = async (): Promise<any> => {
+  return await getContributors();
+};
+
+export const metadata: Metadata = {
+  title: 'Contributors',
+  description: 'Elixir Cloud & AAI contributor developing the solutions.',
+};
+
+const Contributors: FC<any> = async () => {
```
Comment (anuragxxd) [inline src/app/contributors/page.tsx:18]: Instead of `FC<any>` why aren't we using  just `FC` here as type?

File: src/app/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        <Script
+          crossOrigin="anonymous"
+          src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.esm.js"
+          type="module"
+        />
+        <Script
+          crossOrigin="anonymous"
+          noModule
+          src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.js"
+        />
```
Comment (anuragxxd) [inline src/app/layout.tsx:43]: We can replace the icons here too now with something better like https://www.npmjs.com/package/@remixicon/react

File: src/app/overview/page.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { Metadata } from 'next';
+import { FC } from 'react';
+
+export const revalidate = 60;
+
+const fetchOverview = async () => {
+  let data = await getOverview();
+  data = JSON.stringify(data);
+  data = JSON.parse(data);
+  return data;
```
Comment (anuragxxd) [inline src/app/overview/page.tsx:13]: Why stringify & parse?

File: src/app/page.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import ServiceWorkerRegistration from '../components/service-worker-registration';
+import React, { FC } from 'react';
+import HomeComponent from '../components/home';
+
+const HomePage: FC<any> = () => {
+  return (
+    <>
+      <ServiceWorkerRegistration />
```
Comment (anuragxxd) [inline src/app/page.tsx:8]: Shouldn't this be done in layout?

File: src/components/service-worker-registration/index.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+          },
+          (err) => {
+            console.log('Service Worker registration failed: ', err);
+          },
+        );
+      });
+    }
+  }, []);
+  return null;
+}
```
Comment (anuragxxd) [inline src/components/service-worker-registration/index.tsx:23]: We can use the package for this implementation for better support: https://serwist.pages.dev/docs/next

File: src/seo/base-seo.ts [hunk 6]
```diff
@@ -1,6 +1,8 @@
-const seo = {
+import { Metadata } from 'next';
+
+const baseSeo: Metadata = {
+  metadataBase: new URL('https://elixir-cloud-aai-github-io.vercel.app'),
```
Comment (anuragxxd) [inline src/seo/base-seo.ts:4]: This is the correct URL: https://elixir-cloud.dcc.sib.swiss/

File: yarn.lock
Comment (anuragxxd) [inline yarn.lock:1]: You can remove `yarn.lock` here & just rely on `package-lock.json`

File: .sample.env
Comment (anuragxxd) [inline .sample.env:1]: We can keep this file here. Helps to get the overview of `envs`

**Review by anuragxxd — COMMENTED** [review]
Hey @dhararaghu777, good work with this. Have left some comments please check them out. LGTM otherwise.

**Comment by anuragxxd** [comment]
> Wouldn't it make more sense to use elixir's eslint config defines in cloud component? @anuragxxd

I don't think it will as there is config for `turbo` there. Maybe we can if we make it more generic & use remove turbo from there, but it we lot of hassle for now i guess!

File: src/app/layout.tsx (hunk 2 above)
Comment (dhararaghu777) [reply src/app/layout.tsx:43]: we are using react icons

File: src/app/overview/page.tsx (hunk 3 above)
Comment (dhararaghu777) [reply src/app/overview/page.tsx:13]: Resolved

File: src/app/page.tsx (hunk 4 above)
Comment (dhararaghu777) [reply src/app/page.tsx:8]: Resolved

File: src/seo/base-seo.ts (hunk 6 above)
Comment (dhararaghu777) [reply src/seo/base-seo.ts:4]: Resolved

File: .sample.env
Comment (dhararaghu777) [reply .sample.env:1]: Resolved

File: yarn.lock
Comment (dhararaghu777) [reply yarn.lock:1]: Resolved

File: src/components/service-worker-registration/index.tsx (hunk 5 above)
Comment (dhararaghu777) [reply src/components/service-worker-registration/index.tsx:23]: Why can't we use  [Next-pwa](https://github.com/shadowwalker/next-pwa). It is more popular than the above mentioned

Comment (dhararaghu777) [reply src/components/service-worker-registration/index.tsx:23]: I will pick up this next PR

**Review by anuragxxd — APPROVED** [review]

**Pull request merged**